src/docfx/lib/git/GitUtility.cs

```diff
@@ -174,10 +174,30 @@ public static Task PruneWorkTrees(string cwd)
 
         /// <summary>
         /// Retrieve git head version
-        /// TODO: For testing purpose only, move it to test
         /// </summary>
-        public static Task<string> Revision(string cwd, string branch = "HEAD")
-           => ExecuteQuery(cwd, $"rev-parse {branch}");
+        public static unsafe string RevParse(string repoPath, string committish = null)
+        {
+            string result = null;
+
+            if (string.IsNullOrEmpty(committish))
+            {
+                committish = "HEAD";
+            }
+
+            if (git_repository_open(out var repo, repoPath) != 0)
+            {
+                throw new InvalidOperationException($"Not a git repo {repoPath}");
+            }
+
+            if (git_revparse_single(out var reference, repo, committish) == 0)
+            {
+                result = git_object_id(reference)->ToString();
+                git_object_free(reference);
```

Comment:
free `reference` outside `if` block?

Reply:
when `git_revparse_single` fails (returns non zero), reference is an invalid point, you can only free a valid pointer.